### PR TITLE
feat: wire tracing across rpc boundaries

### PR DESCRIPTION
**Pull request by frrist**
This PR wires up tracing across the RPC boundaries of the estuary node and its shuttles. Attached is a trace of a file being added to Estuary via the `/content/add-ipfs` endpoint, the request being made to the shuttle, and the response of the shuttle to estuary indicating the file has been pinned successfully.
tracingestuaryscreenshot

I am unsure if the shuttle nodes are configured to talk with lightstep, if they are this will work out fo the box, otherwise shuttles will need further work to setup that connection. Not sure how to do that..

### CODE REVIEW

File: cmd/estuary-shuttle/main.go [hunk 1]
```diff
@@ -51,8 +53,6 @@ import (
 	"github.com/whyrusleeping/memo"
 )
 
-var Tracer = otel.Tracer("shuttle")
```
Comment (frrist) [inline cmd/estuary-shuttle/main.go:54]: this moved onto the Shuttle struture so they can be viewed independently, i.e. which shuttle produced which trace.

File: cmd/estuary-shuttle/main.go [hunk 2]
```diff
@@ -232,6 +253,14 @@ func main() {
 			return err
 		}
 
+		if cctx.Bool("jaeger-tracing") {
```
Comment (frrist) [inline cmd/estuary-shuttle/main.go:256]: allows estuary/shuttles to send traces to a jaeger endpoint. Disabled by default.

File: cmd/estuary-shuttle/main.go [hunk 3]
```diff
@@ -403,16 +437,16 @@ type chanTrack struct {
 	last *filclient.ChannelState
 }
 
-func (d *Shuttle) RunRpcConnection() error {
```
Comment (frrist) [inline cmd/estuary-shuttle/main.go:406]: consistency, sorry for the extra diff.

File: cmd/estuary-shuttle/rpc.go [hunk 4]
```diff
@@ -11,53 +11,70 @@ import (
...
+	"go.opentelemetry.io/otel/trace"
 	"golang.org/x/xerrors"
 	"gorm.io/gorm"
 )
 
-func (d *Shuttle) handleRpcCmd(cmd *drpc.Command) error {
+func (s *Shuttle) handleRpcCmd(cmd *drpc.Command) error {
 	ctx := context.TODO()
 
+	// If the command contains a trace continue it here.
```
Comment (frrist) [inline cmd/estuary-shuttle/rpc.go:23]: Here is where the shuttle continues the trace if present in the command from Estuary.

File: cmd/estuary-shuttle/rpc.go [hunk 5]
```diff
@@ -315,58 +354,85 @@ func (s *Shuttle) trackTransfer(chanid *datatransfer.ChannelID, dealdbid uint) {
...
 			return
 		}
 
-		d.trackTransfer(chanid, cmd.DealDBID)
+		s.trackTransfer(chanid, cmd.DealDBID)
 
-		if err := d.sendRpcMessage(ctx, &drpc.Message{
-			Op: drpc.OP_TransferStarted,
+		if err := s.sendRpcMessage(ctx, &drpc.Message{
+			TraceCarrier: drpc.NewTraceCarrier(span.SpanContext()),
```
Comment (frrist) [inline cmd/estuary-shuttle/rpc.go:384]: Here is one example of how the shuttle passes the trace context back to estuary.

File: drpc/traceing.go [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+import (
+	"encoding/json"
+
+	"go.opentelemetry.io/otel/trace"
+)
+
+/*
+TraceCarrier is required to Marshal and Unmarshall trace.SpanContext across RPC boundaries since the
+OpenTelemetry spec dictates trace.SpanContext must be read only, and therefore does not implement a
```
Comment (frrist) [inline drpc/traceing.go:11]: Bit annoying, but below is the workaround for this.

File: main.go [hunk 7]
```diff
@@ -33,6 +33,12 @@ import (
 	"gorm.io/gorm/logger"
 )
 
+func init() {
```
Comment (frrist) [inline main.go:36]: allows estuary to be started without a lotus node running on the same machine.

File: shuttle.go [hunk 8]
```diff
@@ -116,7 +117,15 @@ func (cm *ContentManager) registerShuttleConnection(handle string, hello *drpc.H
 var ErrNilParams = fmt.Errorf("shuttle message had nil params")
 
 func (cm *ContentManager) processShuttleMessage(handle string, msg *drpc.Message) error {
-	ctx, span := cm.tracer.Start(context.TODO(), "processShuttleMessage")
+	ctx := context.TODO()
+
+	// if the message contains a trace continue it here.
+	if msg.HasTraceCarrier() {
```
Comment (frrist) [inline shuttle.go:123]: Here is where Estuary will continue a trace if one is present in the message received from the shuttle.

File: cmd/estuary-shuttle/main.go [hunk 9]
```diff
@@ -135,6 +135,27 @@ func main() {
...
+		&cli.BoolFlag{
+			Name:  "jaeger-tracing",
+			Value: false,
+		},
+		&cli.StringFlag{
+			Name:  "jaeger-provider-host",
+			Value: "localhost",
+		},
+		&cli.IntFlag{
+			Name:  "jaeger-provider-port",
```
Comment (whyrusleeping) [inline cmd/estuary-shuttle/main.go:147]: would be nice to combine all these flags into one (or two, to set the sampler ratio separately)

File: cmd/estuary-shuttle/rpc.go [hunk 10]
```diff
@@ -67,14 +84,17 @@ func (d *Shuttle) addPin(ctx context.Context, contid uint, data cid.Cid, user ui
 			log.Errorf("have multiple pins for same content id: %d", contid)
 		}
 		existing := search[0]
+		span.AddEvent("previous found a pin locally", trace.WithAttributes(attribute.Int64("ID", int64(existing.ID))))
 
 		if existing.Failed {
+			span.AddEvent("detected previous pin failure, resend notification")
```
Comment (whyrusleeping) [inline cmd/estuary-shuttle/rpc.go:90]: im wondering how much extra load will be generated by these events...

File: cmd/estuary-shuttle/rpc.go [hunk 11]
```diff
@@ -167,8 +191,9 @@ func (d *Shuttle) handleRpcComputeCommP(ctx context.Context, cmd *drpc.ComputeCo
 		return xerrors.Errorf("result from commp memoizer was of wrong type: %T", res)
 	}
 
-	return d.sendRpcMessage(ctx, &drpc.Message{
-		Op: drpc.OP_CommPComplete,
+	return s.sendRpcMessage(ctx, &drpc.Message{
```
Comment (whyrusleeping) [inline cmd/estuary-shuttle/rpc.go:194]: Maybe we should just set the TraceCarrier in the sendRpcMessage method?

File: main.go [hunk 12]
```diff
@@ -193,6 +199,27 @@ func main() {
...
 		},
+		&cli.BoolFlag{
+			Name:  "jaeger-tracing",
+			Value: false,
+		},
+		&cli.StringFlag{
+			Name:  "jaeger-provider-host",
+			Value: "localhost",
+		},
+		&cli.IntFlag{
```
Comment (whyrusleeping) [inline main.go:210]: same comment about making this fewer flags

**Review by whyrusleeping — CHANGES_REQUESTED** [review]
A few small changes requested, otherwise looks great, thanks @frrist !

**Review by whyrusleeping — APPROVED** [review]
lgtm

**Pull request merged**